### ISSUE TITLE
hardlink_master bug

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#7175]
BUG: borg 1.x does not set `item.hardlink_master = <boolean>` for fifos, block and char devices (but only for regular files).

note: it is not needed to be set for directories and symlinks, because they are not considered hard-linkable for borg1.

the borg2 upgrader code used for `borg transfer` used this up to 2.0.0b4:
```
    def borg1_hardlink_master(self, item):  # legacy
        return item.get("hardlink_master", True) and "source" not in item and self.borg1_hardlinkable(item.mode)
```
So:
- not hardlinked regular files: had `item.hardlink_master = False` set, ok
- hardlink master regular files: had `item.hardlink_master = True` set, ok
- fifos, block and char devices never had hard `.hardlink_master` set (neither True, nor False), so they **defaulted to True, which would be rather unpretty and unneeded in borg2 archives, because they all would get `item.hlid` set for no reason.** BUG
- hardlink slaves (all hardlinkable types) have `.source` set (although they are not symlinks), ok

**Comment by ThomasWaldmann** [comment]
Fix for borg 1.x:

Store hardlink_master also for fifos, block and char devices (see code for regular files).

Fix for borg2:
```
    def borg1_hardlink_master(self, item):  # legacy
        return item.get("hardlink_master", False) and "source" not in item and self.borg1_hardlinkable(item.mode)
```

For old borg 1.x archives with the bug, this will extract originally hardlinked fifos, block and char devices as individual (not hardlinked) files. Guess this would rarely be a problem, who hardlinks such stuff anyway?

For newer borg 1.x archives (made after fixing this), it will behave correctly for hardlinks of all file types.

**Comment by ThomasWaldmann** [comment]
Fix for master branch applied by c343673.

**Comment by ThomasWaldmann** [comment]
Fixed by #7412.
